### PR TITLE
building: Add --include-syslibs option.

### PR DESCRIPTION
**Pull request by DrDaveD**
This implements the feature requested in issue #2732.  It allows
excluding the majority of shared system libraries from the packaged
bundle, except for those explicitly requested by this option.

### CODE REVIEW

**Comment by DrDaveD** [comment]
On my system (CentOS 7) and my application I have to use `--include-syslibs ffi-` in order to include two libraries which seem to be have names generated with hashes, specifically `libffi-c643fa1a.so.6.0.4` and `libffi-ae16d830.so.6.0.4`, or it crashes at runtime.  I don't know why they're needed, and I'm not sure how universally they are needed so for now at least I left `ffi-` out of being included by default.  I wouldn't mind putting it in to make it easier for people to use if people think it's a good idea.

**Comment by bwoodsend** [comment]
> On my system (CentOS 7) and my application I have to use --include-syslibs ffi- in order to include two libraries which seem to be have names generated with hashes, specifically libffi-c643fa1a.so.6.0.4 and libffi-ae16d830.so.6.0.4, or it crashes at runtime.

Do you not have a more readable named symlink somewhere in one in one of your distribution's `/lib/` folders? That's the usual Unix way - have the lib itself be named something hideous then have a nice `libffi.so` symlink to make it easier to find.

**Comment by bwoodsend** [comment]
This is looking good.

>  If not set or an empty list, all will be included.

I'm not too sure I like this behaviour. It would inhibit collecting no libraries at all. Perhaps we could use a comma separated list argument rather than one you can respecify so that `--include-syslibs=''` means collect none, `--include-syslibs='z,ffi'` means collected libz and libffi and then not using the argument at all still means collect everything?

I'm also not keen on the `sys` bit of the name `--include-syslibs`. *System libraries* to me means  the really core stuff from the posix specification such as libc or libm which we actively avoid including anyway. The libraries you're targeting aren't really system libraries - they're just not part of your Python installation. That said, I can't think of anything better to call it.

**Comment by rokm** [comment]
I honestly don't see much value of having this implemented as part of Analysis, as opposed to a separate function that post-processes Analysis object and filters its `binaries` and `datas`. In the latter case, everyone can tailor the behavior exactly to their use case (ignore everything that comes from /usr (except python lib and modules)), ignore just some binaries, etc.). Granted, it is not out-of-the-box, but then again, filtering like this is an advanced requirement (and its exact requirements depend on the use case). To alleviate this, we could perhaps provide an implementation for users to import and use in their spec as post-processing step.

In either case (built-in or external post-processing function), the filtering logic should be reconsidered - if you are trying to ignore binaries coming from system (e.g., /usr/lib{,64}, /lib{,64}), the filtering should probably be done based on the source path, not destination. After all, collected python packages may provide their own copy of `libsomething.so`, which will be collected into top-level directory and which you don't want to filter out (for example, `libopenblasp-r0-5bebc122.3.13.dev.so` that `numpy` wheel brings in; or `libffi-c643fa1a.so.6.0.4` that `cffi` wheel brings in). And package directories with datafiles and/or binary extensions may have their name starting with `lib` as well...

Also, if you don't care about shared libraries from the system, you probably should not care about data files, either (although then you might run afoul of runtime hooks)...

**Comment by DrDaveD** [comment]
> > On my system (CentOS 7) and my application I have to use --include-syslibs ffi- in order to include two libraries which seem to be have names generated with hashes, specifically libffi-c643fa1a.so.6.0.4 and libffi-ae16d830.so.6.0.4, or it crashes at runtime.
> 
> Do you not have a more readable named symlink somewhere in one in one of your distribution's `/lib/` folders? That's the usual Unix way - have the lib itself be named something hideous then have a nice `libffi.so` symlink to make it easier to find.

There is also a libffi.so with the kind of symlink you're talking about but it ends up at libffi.so.6.0.1.  These oddly named ones don't exist anywhere else on my system (`locate` doesn't find them) so I don't know where they are coming from.

**Comment by DrDaveD** [comment]
Oh, I see @rokm said they come from the `cffi` python package.

**Comment by DrDaveD** [comment]
> > If not set or an empty list, all will be included.
> 
> I'm not too sure I like this behaviour. It would inhibit collecting no libraries at all. Perhaps we could use a comma separated list argument rather than one you can respecify so that `--include-syslibs=''` means collect none, `--include-syslibs='z,ffi'` means collected libz and libffi and then not using the argument at all still means collect everything?
> 
> I'm also not keen on the `sys` bit of the name `--include-syslibs`. _System libraries_ to me means the really core stuff from the posix specification such as libc or libm which we actively avoid including anyway. The libraries you're targeting aren't really system libraries - they're just not part of your Python installation. That said, I can't think of anything better to call it.

Oops, I actually called the option `--include-syslib` but made the wrong commit title to say it was `--include-syslibs`.  I'd be happy to change it to `--include-syslibs` and make it be a comma-separated list instead of multiple `--include-syslib`, that is friendlier anyway.  I can't think of a better name either.  To me a system lib is anything from the operating system, not just the most basic ones.

**Comment by DrDaveD** [comment]
> I honestly don't see much value of having this implemented as part of Analysis, as opposed to a separate function that post-processes Analysis object and filters its `binaries` and `datas` (similarly to `MERGE`). In the latter case, everyone can tailor the behavior exactly to their use case (ignore everything that comes from /usr (except python lib and modules)), ignore just some binaries, etc.). Granted, it is not out-of-the-box, but then again, filtering like this is an advanced requirement (and its exact requirements depend on the use case). To alleviate this, we could perhaps provide an implementation for users to import and use in their spec as post-processing step.

Maybe I'm not following you completely, but that sounds like it's still going to need a hack to the spec file, making it less easy to use.  The idea behind this is to make it easier for people to use.

> In either case (built-in or external post-processing function), the filtering logic should be reconsidered - if you are trying to ignore binaries coming from system (e.g., /usr/lib{,64}, /lib{,64}), the filtering should probably be done based on the source path, not destination. After all, collected python packages may provide their own copy of `libsomething.so`, which will be collected into top-level directory and which you don't want to filter out (for example, `libopenblasp-r0-5bebc122.3.13.dev.so` that `numpy` wheel brings in; or `libffi-c643fa1a.so.6.0.4` that `cffi` wheel brings in). And package directories with datafiles and/or binary extensions may have their name starting with `lib` as well...

That's a good idea.  I'll change it to look at the full path of the source and make sure it only operates on those in paths starting with `/lib*/lib` or `/usr/lib*/lib`.

> Also, if you don't care about shared libraries from the system, you probably should not care about data files, either (although then you might run afoul of runtime hooks)...

I'm not sure what you mean about data files.  There aren't any system data files, are there?

**Comment by rokm** [comment]
> > I honestly don't see much value of having this implemented as part of Analysis, as opposed to a separate function that post-processes Analysis object and filters its `binaries` and `datas` (similarly to `MERGE`). In the latter case, everyone can tailor the behavior exactly to their use case (ignore everything that comes from /usr (except python lib and modules)), ignore just some binaries, etc.). Granted, it is not out-of-the-box, but then again, filtering like this is an advanced requirement (and its exact requirements depend on the use case). To alleviate this, we could perhaps provide an implementation for users to import and use in their spec as post-processing step.
> 
> Maybe I'm not following you completely, but that sounds like it's still going to need a hack to the spec file, making it less easy to use. The idea behind this is to make it easier for people to use.

I know. And I'm rather hesitant to make it easier for users to shoot themselves in the foot. Building pyinstaller-frozen applications on linux is enough of a minefield as it is.

Also, I think for any reasonably sized and complex application, building via .spec should be preferred approach anyway...

But I also see where you're coming from, so if other developers think this is a good idea, then I'll keep my concerns about the idea in general to myself.

> 
> > In either case (built-in or external post-processing function), the filtering logic should be reconsidered - if you are trying to ignore binaries coming from system (e.g., /usr/lib{,64}, /lib{,64}), the filtering should probably be done based on the source path, not destination. After all, collected python packages may provide their own copy of `libsomething.so`, which will be collected into top-level directory and which you don't want to filter out (for example, `libopenblasp-r0-5bebc122.3.13.dev.so` that `numpy` wheel brings in; or `libffi-c643fa1a.so.6.0.4` that `cffi` wheel brings in). And package directories with datafiles and/or binary extensions may have their name starting with `lib` as well...
> 
> That's a good idea. I'll change it to look at the full path of the source and make sure it only operates on those in paths starting with `/lib*/lib` or `/usr/lib*/lib`.
> 

Just remember to exclude path to python library and its extension modules, as well as anything coming from site-packages.

> > Also, if you don't care about shared libraries from the system, you probably should not care about data files, either (although then you might run afoul of runtime hooks)...
> 
> I'm not sure what you mean about data files. There aren't any system data files, are there?

Resources from `/usr/share` might end up being collected depending on used packages and libraries. For example, freezing a `gtk3` application ends up collecting icons, themes, etc. So if one wanted to avoid collecting `gtk3` (and underlying) libraries, they would probably want to avoid collecting those resources as well.

**Comment by rokm** [comment]
> Maybe I'm not following you completely, but that sounds like it's still going to need a hack to the spec file, making it less easy to use. The idea behind this is to make it easier for people to use.

Actually, the only thing here that makes things easier to use here is the command-line switch. And that could be used to inject the post-processing function call into the .spec file (similarly to how we do it for --collect-* switches).

It's a bit more work to implement, but I would find it preferable to having yet another option added to Analysis and having the filter function built in there.

**Comment by bwoodsend** [comment]
I'm with @rokm on it being spec file only. Either that or we could do what pytest do with some of their less official options and name it `--include-syslibs-and-revoke-all-rights-to-community-support-and-access-to-the-pyinstaller-issue-tracker`.

**Comment by DrDaveD** [comment]
For now I have updated the commit/PR to use a comma-separated list for `--include-syslibs` and to filter based on both the source being under `/lib*` or `/usr/lib*` and the source and destination starting with `lib` but not `lib-dynload`.

I'll wait to hear if others think it should be changed to be done via postprocessing in the spec file instead of in the analysis step.

Thanks for the explanation about the data files, but I'm not sure what could be done about them.

(By the way, earlier I was talking about two libffi-XXXX files, but it turned out I somehow had some leftover files from a previous install that didn't get cleaned out by the packaging system.  The actual package only includes one.)

**Review by bwoodsend — CHANGES_REQUESTED** [review]
And at some point this will need a [news entry](https://pyinstaller.readthedocs.io/en/stable/development/changelog-entries.html) but I think we should work out what we're going to do with this feature first.

**Comment by bwoodsend** [comment]
For data files, I can't think of anything other than Gtk which has them and Gtk already has special handling to restrict the data files it uses.

**Comment by bwoodsend** [comment]
> But I also see where you're coming from, so if other developers think this is a good idea, then I'll keep my concerns about the idea in general to myself.

Personally, I'm ok with this. At least to eliminate libraries like `libssl` which we really should not be bundling. There was also an issue about LGPL libraries like readline being sucked in which makes any closed source application technically breaching copyright. Personally I think that exlcuding them from the application only to install them with your package manager is just jumping through hoops but this would hopefully appease them too.

File: PyInstaller/building/build_main.py [hunk 1]
```diff
@@ -536,6 +539,22 @@ def assemble(self):
...
+                if path.startswith('lib-dynload/') or path.startswith('libpython'):
+                    return True
+                for syslib in self.syslibs:
+                    if syslib is None:
+                        return True
+                    if path.startswith('lib' + syslib):
+                        return True
+                return False
+            self.binaries = list(filter(_should_include_binary, self.binaries))
+
```
[the remark this answers is not in the document]
Comment (DrDaveD) [reply PyInstaller/building/build_main.py]: Done.  It's not clear to me where exactly the unit test should go though.

**Comment by DrDaveD** [comment]
> I'm with @rokm on it being spec file only. Either that or we could do what pytest do with some of their less official options and name it `--include-syslibs-and-revoke-all-rights-to-community-support-and-access-to-the-pyinstaller-issue-tracker`. Otherwise we'll just get hotheads thinking that it's a magic switch to shrink your Linux builds and apply it without reading about the consequences.

I didn't notice this post until just now.  Are you saying that you also would like this to be moved from where it is now?  I got the impression from your later messages that you were ok with it where it was, but maybe I misinterpret.  Should I keep the option name as it is, and insert it into post-processing similar to the way that the `--collect-*` options do?

**Comment by bwoodsend** [comment]
I'd prefer that it wasn't a command line option. Rather you have to edit your spec file to use this feature.

**Comment by DrDaveD** [comment]
I hardly see the point in that, since people can already do that now.  You're saying that a function will be provided that they can call?  I guess that would be some help.  What would the code in the spec file that they would need to add themselves look like?

**Comment by rokm** [comment]
> I hardly see the point in that, since people can already do that now. You're saying that a function will be provided that they can call? I guess that would be some help. 

Yes, the idea is to provide base filtering function.

> What would the code in the spec file that they would need to add themselves look like?

```
...
a = Analysis(...)
exclude_system_libraries(a, list_of_exceptions=['*blas*', '*ffi*', 'libfreetype*'])  # filters a.binaries to exclude system libraries
pyz = PYZ(...)
...
```

**Comment by bwoodsend** [comment]
Just another option in the spec file.

```python
a = Analysis(
    ...,
    syslibs=["ffi", "dl"],
    ...,
)

**Comment by rokm** [comment]
> Just another option in the spec file.
> 
> ```python
> a = Analysis(
>     ...,
>     syslibs=["ffi", "dl"],
>     ...,
> )
> ```

I'd rather not; having `syslibs` argument in `Anaylsis` can be very misleading, as it offers no hint that it implies (or is even related to) exclusion... (same goes for the proposed `--include-syslibs` command-line switch).

**Comment by bwoodsend** [comment]
Oh, I didn't see [this one](https://github.com/pyinstaller/pyinstaller/pull/6016#issuecomment-881011998). I prefer that to my suggestion.

**Comment by DrDaveD** [comment]
Well, that's a whole different PR.  I'll close this one and make another later.

**Pull request closed without merging**